// m2m-dali/lib/dali-sequences.js
//dali-sequences.js
//Author Alex Pressl for imitrix GmbH
var async = require('async');
var Serial = require ('serialport').SerialPort;
var serialPort;


exports.seq = {};

exports.seq.Initialize = [
	['sendcmd',0xFF, 0x00, 'set lamp OFF'],
	['sendcmd',0xFF, 0x01, 'inc dim fade 200ms'],
	['sendcmd',0xFF, 0x21, 'store current light level in DTR'],
	['sendcmd',0xFF, 0x2B, 'set minimum level		with DTR'],
	['sendcmd',0xFF, 0x2D, 'set power on level 	with DTR'],

	['sendcmd',0xFF, 0x05, 'set arc level to maximum (full ON)'],
	['sendcmd',0xFF, 0x02, 'dec dim fase 200ms'],
	['sendcmd',0xFF, 0x21, 'store current light level in DTR'],
	['sendcmd',0xFF, 0x2A, 'set maximum level		with DTR'],

	['sendcmd',0xFF, 0x02, 'dec dim fase 200ms'],
	['sendcmd',0xFF, 0x02, 'dec dim fase 200ms'],
	['sendcmd',0xFF, 0x02, 'dec dim fase 200ms'],
	['sendcmd',0xFF, 0x02, 'dec dim fase 200ms'],
	['sendcmd',0xFF, 0x21, 'store current light level in DTR'],
	['sendcmd',0xFF, 0x2C, 'set failure level		with DTR'],

	['sendcmd',0xFF, 0x06, 'set arc level to minimum value'],
	['sendcmd',0xFF, 0x05, 'set arc level to maximum (full ON)'],
];

exports.seq.ForceMin = [
	['sendcmd',0xFF, 0x00, 'set lamp OFF'],
	['sendcmd',0xFF, 0x21, 'store current light level in DTR'],
	['sendcmd',0xFF, 0x2B, 'set minimum level		with DTR'],
	['sendcmd',0xFF, 0x2B, 'set minimum level		with DTR'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x21, 'store current light level in DTR'],
	['sendcmd',0xFF, 0x2D, 'set power on level 	with DTR'],
	['sendcmd',0xFF, 0x2D, 'set power on level 	with DTR'],


	['sendcmd',0xFF, 0x05, 'set arc level to maximum (full ON)'],
	['sendcmd',0xFF, 0x06, 'set arc level to minimum value'],
	['sendcmd',0xFF, 0x05, 'set arc level to maximum (full ON)'],
	['sendcmd',0xFF, 0x06, 'set arc level to minimum value'],
];

exports.seq.Check = [
	['sendcmd',0xFF, 0x06, 'set arc level to minimum value'],
	['sendcmd',0xFF, 0x9A, 'ret physical level minimum value'],
	['sendcmd',0xFF, 0xA2, 'ret minimum level  value'],
	['sendcmd',0xFF, 0xA0, 'ret current level  value'],
	['sendcmd',0xFF, 0x21, 'store current light level in DTR'],
	['sendcmd',0xFF, 0x2D, 'set power on level 	with DTR'],
	['sendcmd',0xFF, 0x00, 'set lamp OFF'],
	['sendcmd',0xFF, 0x06, 'set arc level to minimum value'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0xA0, 'ret current level  value'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0xA0, 'ret current level  value'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0xA0, 'ret current level  value'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0xA0, 'ret current level  value'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0xA0, 'ret current level  value'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0xA0, 'ret current level  value'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x06, 'set arc level to minimum value'],
	['sendcmd',0xFF, 0x05, 'set arc level to maximum (full ON)'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x05, 'set arc level to maximum (full ON)'],
	['sendcmd',0xFF, 0x04, 'dec2 arc level'],
	['sendcmd',0xFF, 0x04, 'dec2 arc level'],
	['sendcmd',0xFF, 0x04, 'dec2 arc level'],
	['sendcmd',0xFF, 0x04, 'dec2 arc level'],
	['sendcmd',0xFF, 0x04, 'dec2 arc level'],
	['sendcmd',0xFF, 0x04, 'dec2 arc level'],
	['sendcmd',0xFF, 0x04, 'dec2 arc level'],
	['sendcmd',0xFF, 0x04, 'dec2 arc level'],
	['sendcmd',0xFF, 0x04, 'dec2 arc level'],
	['sendcmd',0xFF, 0x04, 'dec2 arc level'],
	['sendcmd',0xFF, 0x04, 'dec2 arc level'],
	['sendcmd',0xFF, 0x04, 'dec2 arc level'],
	['sendcmd',0xFF, 0x04, 'dec2 arc level'],
	['sendcmd',0xFF, 0x04, 'dec2 arc level'],
	['sendcmd',0xFF, 0x04, 'dec2 arc level'],
	['sendcmd',0xFF, 0x04, 'dec2 arc level'],
	['sendcmd',0xFF, 0x04, 'dec2 arc level'],
	['sendcmd',0xFF, 0x04, 'dec2 arc level'],
	['sendcmd',0xFF, 0x04, 'dec2 arc level'],
	['sendcmd',0xFF, 0x04, 'dec2 arc level'],
	['sendcmd',0xFF, 0x06, 'set arc level to minimum value'],

	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x03, 'inc2 arc level'],
	['sendcmd',0xFF, 0x05, 'set arc level to maximum (full ON)'],
];


exports.seq.Test = [
	['sendcmd',0xFF, 0x00, 'set lamp OFF'],
	['sendcmd',0xFF, 0x06, 'set arc level to minimum value'],

	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],
	['sendcmd',0xFF, 0x08, 'inc arc level'],

	['sendcmd',0xFF, 0x05, 'set arc level to maximum (full ON)'],
	['sendcmd',0xFF, 0x05, 'set arc level to maximum (full ON)'],
	['sendcmd',0xFF, 0x05, 'set arc level to maximum (full ON)'],
	['sendcmd',0xFF, 0x05, 'set arc level to maximum (full ON)'],
	['sendcmd',0xFF, 0x05, 'set arc level to maximum (full ON)'],

	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],
	['sendcmd',0xFF, 0x07, 'dec arc level'],

	['sendcmd',0xFF, 0x06, 'set arc level to minimum value'],
	['sendcmd',0xFF, 0x06, 'set arc level to minimum value'],
	['sendcmd',0xFF, 0x00, 'set lamp OFF']
];
exports.seq.DimUp = [
	['sendcmd',0xFF, 0x08, 'inc arc level']
];
exports.seq.DimDown = [
	['sendcmd',0xFF, 0x07, 'dec arc level']
];
exports.seq.On = [

	['sendcmd',0xFF, 0x05, 'set arc level to maximum (full ON)']
];

exports.seq.Off = [

	['sendcmd',0xFF, 0x00, 'set lamp OFF']
];

exports.seq.Min = [

	['sendcmd',0xFF, 0x05, 'set arc level to maximum (full ON)'],

	['sendcmd',0xFF, 0x02, 'dec dim fase 200ms'],
	['sendcmd',0xFF, 0x02, 'dec dim fase 200ms'],
	['sendcmd',0xFF, 0x02, 'dec dim fase 200ms'],
	['sendcmd',0xFF, 0x02, 'dec dim fase 200ms'],
	['sendcmd',0xFF, 0x02, 'dec dim fase 200ms'],
	['sendcmd',0xFF, 0x02, 'dec dim fase 200ms'],
	['sendcmd',0xFF, 0x02, 'dec dim fase 200ms'],
	['sendcmd',0xFF, 0x06, 'set arc level to minimum value'],
];
var myPlugin;
var cache = '';
var serialportOpen = false;

exports.initSeq = function (plugin, serialdev) {
	myPlugin = plugin;

	if (process.argv.length <= 2) { 
	    try {
		    serialPort = new Serial (serialdev, {baudrate:115200}, true);
	    } catch (err) {
			myPlugin.logger.log('error','Serial DALI port create failed: ' + err);
	    }
		serialPort.on('open', function(data) {
			myPlugin.logger.log('info','Serial DALI port open: ' + data);
			serialportOpen = true;
		});
		serialPort.on('data', function(data) {
			myPlugin.logger.log('info','Serial DALI data received: ' + data);
		});
	}

}

exports.executeSeq = function (seq) {
    if (serialportOpen==true) {
		async.eachSeries (seq, sendcmd, executeSeqComplete);
	} else {
		myPlugin.logger.log('error','serialport closed');

	}
}

function executeSeqComplete(err) {
	//serialPort.close();
	if (err)
		myPlugin.logger.log('error','executeSeq Failed');
}

function sendcmd(opcode, callback) {
	var adr = (("0" + opcode[1].toString(16)).substr(-2)).toUpperCase();
	var pay = (("0" + opcode[2].toString(16)).substr(-2)).toUpperCase();
    //myPlugin.logger.log('info','Send: '+ opcode[0]+' '+adr+' '+pay+' >>'+opcode[3]);
    sendSeq(opcode);

	var result = 'ok';
    setTimeout(function(){
      callback(null, result);
    }, 400);	
}

function sendSeq(msg) {
	{
		{
            //myPlugin.logger.log('info','Serial port is connected and opened');
            var a = parseInt(msg[1]);
            var adr = (("0" + a.toString(16)).substr(-2)).toUpperCase();
            c = parseInt(msg[2]);
            var cmd = (("0" + c.toString(16)).substr(-2)).toUpperCase();
            var frame = "\x02DALI:"+adr+cmd+"\x03";
            try {
	            serialPort.write(frame, function(err, results) {
	                if (err) {
	                    myPlugin.logger.log('error', 'Error Write 1'+frame+' to DALI serialport : '+ err);
	                }
	                myPlugin.logger.log('info','DALI.BUS:'+frame+' bytes: ' + results);
	                
	            });
            } catch (err) {
                myPlugin.logger.log('error','exception serialport.write '+err);
            }
        } 
    }
}
